### PR TITLE
Fix rm rf v2

### PR DESCRIPTION
**Pull request by jsynacek**

### CODE REVIEW

**Comment by jsynacek** [comment]
Ok, waiting for Zbyszek before splitting the commit.

**Comment by martinpitt** [comment]
@jsynacek semaphore fails on `FAIL: test-unit-name`, that should reproduce locally with `make check`.

**Comment by jsynacek** [comment]
v2.

**Comment by keszybz** [comment]
I think @martinpitt's point about autogenerated paths is very valid. We shouldn't disallow things unless there's a clear reason. So I think the restrictions which are left after patch 1/2 are OK.

It certainly doesn't help it isn't defined anywhere what `path_is_safe` means and for what it is "safe". It seems that every call point makes some implicit assumptions... Going over the callers, allowing `//` and `/./` seems safe. Some callers call `path_kill_slashes` anyway before calling `path_is_safe`, which makes the check for `//` moot. Others check that `path_is_absolute`, so the check for `./` is moot.

2/2 looks OK too.

**Comment by evverx** [comment]
Does anybody know what precisely `path_is_safe` checks? Previously, the condition `path_is_safe == true` guaranteed that `path` is normalised. What does it mean now? I still can remove everything with `/etc/..////`.

**Comment by keszybz** [comment]
Well, `/../` is still forbidden, so you can't.

See my comment — I don't think anyone knows what it's supposed to mean exactly. Do any of the callers actually care if the path is normalized?

**Comment by evverx** [comment]
> Well, /../ is still forbidden, so you can't.

Right. Sorry.

> Do any of the callers actually care if the path is normalized?

Hm, I don't know. If nothing cares why do we need to call `path_is_safe` at all?

**Comment by evverx** [comment]
I've just updated `test-path-util.c`
```diff
diff --git a/src/test/test-path-util.c b/src/test/test-path-util.c
index 22df20a..ad30794 100644
--- a/src/test/test-path-util.c
+++ b/src/test/test-path-util.c
@@ -38,6 +38,12 @@
                 assert_se(path_equal(b, a) == !result);   \
         }

+static void test_path_is_safe(void) {
+        assert_se(path_is_safe("///") == true);
+        assert_se(path_is_safe("/etc/..////") == false);
+        assert_se(path_is_safe("/etc/.////") == true);
+}
+
 static void test_path(void) {
         _cleanup_close_ int fd = -1;

@@ -563,6 +569,7 @@ int main(int argc, char **argv) {
         test_file_in_same_dir();
         test_filename_is_valid();
         test_hidden_or_backup_file();
+        test_path_is_safe();

         test_systemd_installation_has_version(argv[1]); /* NULL is OK */

```

It looks strange. Sorry, but I think `path_is_safe` doesn't make sense now. I don't understand why `/etc/.///` is safer than `/etc/..///`.

**Comment by keszybz** [comment]
The way I understand it, repeated slashes and `/./` as a component are noise, but harmless. But `/../` allows one to go "outside", and is a problem.

**Comment by evverx** [comment]
> repeated slashes and /./ as a component are noise, but harmless.

It depends on a caller. `rm_rf` shouldn't allow `///`. This is why `rm_rf` calls `path_equal(path, "/")`.

> But /../ allows one to go "outside", and is a problem.

Well, it depends on a caller too. Going outside is not always a problem.

**Comment by evverx** [comment]
Hm, `rm_rf` shouldn't allow `///.///` either. But
```diff
+        assert_se(path_is_safe("///.//") == true);
+        assert_se(path_equal("///.//", "/") == false);
```
passes.

I've already asked https://github.com/systemd/systemd/pull/5653#discussion_r108469736, but why don't we try to normalise `path`?

**Comment by martinpitt** [comment]
I don't understand why `///` or `//.//` should not be allowed -- these are just aliases for the root file system, and if you rm those or `/` directly makes no difference -- relying on some obscure pattern heuristic is not the point here, if you try and remove the root file system you just get what you are asking for..

**Comment by evverx** [comment]
>  if you try and remove the root file system you just get what you are asking for

@martinpitt , sorry, I am a bit confused. What is the point of this PR and https://github.com/systemd/systemd/pull/5653?

**Comment by martinpitt** [comment]
The point, as far as I understood it, was to avoid path traversal with an implied `..`; particularly this happened with something like `/etc/.*` which looks innocuous at first sight (as it just looks like "remove all hidden files"), but is disastrous as `.*` also expands to `..`.

File: src/basic/path-util.c [hunk 1]
```diff
@@ -727,10 +727,7 @@ bool path_is_safe(const char *p) {
                 return false;
 
         /* The following two checks are not really dangerous, but hey, they still are confusing */
-        if (startswith(p, "./") || endswith(p, "/.") || strstr(p, "/./"))
-                return false;
-
-        if (strstr(p, "//"))
+        if (startswith(p, "./") || endswith(p, "/."))
```
Comment (poettering) [inline src/basic/path-util.c:730]: uh? why this change? as the comment says, we check here not only for things that are actively dangerous, but also confusing. I am not seeing why "/./" would ever be a good thing to permit here...

**Comment by poettering** [comment]
I really don't understand why relaxing path_is_safe() would be a good idea. As the comments in the function clarify, we check for "confusing" paths here too. The idea really is that we consider anything "unsafe" that is not a properly build, normalized path. Or to say this differently: it's supposed to detect paths that are the result of "dirty" path concatenations...

**Comment by martinpitt** [comment]
I'm mostly concerned about backwards compatibility with existing `tmpfiles.d` snippets here. As [explained above](https://github.com/systemd/systemd/pull/5655#pullrequestreview-29663412) double slashes are a frequent result of producing paths in code robustly, and I see no justification why these should be not  allowed -- there is no case where they would be unsafe.

**Comment by poettering** [comment]
> I'm mostly concerned about backwards compatibility with existing tmpfiles.d snippets here. As explained above double slashes are a frequent result of producing paths in code robustly, and I see no justification why these should be not allowed -- there is no case where they would be unsafe.

I can agree to that, but my suggestion would be to first call path_kill_slashes() on all parsed strings to remove duplicate slashes.

**Comment by martinpitt** [comment]
Dropping the first patch (changing `path_is_safe()` and instead changing the second to call `path_kill_slashes()` first sounds good to me too.

However, long-term I still wonder why `path_is_safe()` should be so picky about it if we basically always call `path_kill_slashes()` beforehand anyway..

**Comment by boucman** [comment]
maybe rename `path_is_safe` to something like `path_is_readable` ? since that seems to be the real point of that function

**Comment by zackw** [comment]
> if you try and remove the root file system you just get what you are asking for.

Note that `rm` (the utility) will refuse to remove the root file system.  See https://github.com/coreutils/coreutils/blob/e7a2580b96370da03c4d3553ccdf4ee66a14c6a4/src/remove.c#L453  Commentary there implies that this is a POSIX requirement.

**Comment by Rudd-O** [comment]
It seems to me that the original bug:

```
 R! /foo/.* - - - - -
```

is that whatever glob function is used to expand `/foo/.*` is incorrectly matching on the dot dot directory entry, when it should not be doing that in the first place.

Agree that this particular PR does help, but **the real bug has not been fixed**.

**Comment by Rudd-O** [comment]
FWIW

```
[user@engineering ansible]$ python
Python 2.7.13 (default, Jan 12 2017, 17:59:37) 
[GCC 6.3.1 20161221 (Red Hat 6.3.1-1)] on linux2
Type "help", "copyright", "credits" or "license" for more information.
>>> import glob
>>> glob.glob(".*")
[]
>>> glob.glob("/tmp/.*")
['/tmp/.esd-1000', '/tmp/.X0-lock', '/tmp/.Test-unix', '/tmp/.font-unix', '/tmp/.XIM-unix', '/tmp/.ICE-unix', '/tmp/.X11-unix']
>>> 
```

**Comment by Saruspete** [comment]
Agreed with @Rudd-O , the only bug to be fixed is in the glob & recursion.  You only have to skip the "." and ".." entries. 

Do not add restrictions on valid use-cases, like "//" or "/abs/path/to/bin/../var/tmp".
the "./" can also be a valid case for entries generated with tools like "find".

**Comment by Rudd-O** [comment]
Correct.  The initial fix that went in, is incorrect, and this fix, is also superstitious.  The right fix is to (a) comply with POSIX rm, (b) prevent globbing from snarfing dot and dot dot, as it ought to be in this scenario.   Even Python glob gets it right.  We can get it right too.

On April 20, 2017 5:53:11 PM GMT+02:00, Adrien Mahieux <notifications@github.com> wrote:
>Agreed with @Rudd-O , the only bug to be fixed is in the glob &
>recursion.  You only have to skip the "." and ".." entries. 
>
>Do not add restrictions on valid use-cases, like "//" or
>"/abs/path/to/bin/../var/tmp".
>the "./" can also be a valid case for entries generated with tools like
>"find".
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/systemd/systemd/pull/5655#issuecomment-295788912

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by ghost** [comment]
> Even Python glob gets it right

LOL

From OpenBSD libc glob.c:
```
 * Copyright (c) 1989, 1993
 *	The Regents of the University of California.  All rights reserved.
 *
 * This code is derived from software contributed to Berkeley by
 * Guido van Rossum.
```

**Comment by Rudd-O** [comment]
And now we know why :-D

On April 23, 2017 5:48:11 PM GMT+02:00, Pavan Maddamsetti <notifications@github.com> wrote:
>> Even Python glob gets it right
>
>LOL
>
>From OpenBSD libc glob.c:
>```
> * Copyright (c) 1989, 1993
> *	The Regents of the University of California.  All rights reserved.
> *
> * This code is derived from software contributed to Berkeley by
> * Guido van Rossum.
>```
>
>-- 
>You are receiving this because you were mentioned.
>Reply to this email directly or view it on GitHub:
>https://github.com/systemd/systemd/pull/5655#issuecomment-296452392

-- 
Sent from my Android device with K-9 Mail. Please excuse my brevity.

**Comment by poettering** [comment]
@Rudd-O well, we use glibc glob(), which will expand to ".." if needed, and so will bash, as you can easily see:

```
echo ./.*
./. ./.. ./.deps ./.dirstamp ./.libs
```

Moreover, coreutils "rm" command line tool doesn't actually do any globbing, it's bash which does that and just passes the resolved strins to "rm". It's "rm" then which filters stuff ending in some specific suffixes, regardless whether those suffixes where the result of globbing or not.

I must say I find the "rm" logic pretty weird, as it checks for suffixes only, but globbing can be anywhere in the path...

That all said, I think it would be wise for us, if we'd add a wrapper for glibc glob() that filters out "." and ".." from any expansion, and that we'd use in all our own cases. We can do this relatively nicely through the GLOB_ALTDIRFUNC logic.

**Comment by poettering** [comment]
I posted #5792 about that now, to keep track.

**Comment by Rudd-O** [comment]
@poettering if you specify a `dot star` after a slash, yes, then `glob()` will match hidden dirs.  But you *have* to specify the `dot` before the `slash`.  The original bug, according to the reproducer posted there, was about `...../*` matching `dot dot`, which is not correct, not even by glibc `glob()`.

EDIT: no, Lennart is right and I am wrong, I misremembered the reproducer.  `/tmp/.*` is a legit glob expression that will match `dot dot`.
